service/history/workflow/mutable_state_impl.go

```diff
@@ -2358,6 +2361,69 @@ func (ms *MutableStateImpl) addResetPointFromCompletion(
 	return true
 }
 
+// processBuildIdRedirect validates build ID for the task being dispatched and applies possible redirect to
+// mutable state.
+// If a valid redirect is happening, assigned build ID of the wf will be updated and all scheduled but not
+// started tasks will be rescheduled to be put on the matching queue of the right build ID.
+// If the given versioning stamp and redirect info is not valid based on the wf's assigned build ID
+// InvalidDispatchBuildId error will be returned.
+func (ms *MutableStateImpl) processBuildIdRedirect(
+	startingTaskScheduledEventId int64,
+	versioningStamp *commonpb.WorkerVersionStamp,
+	redirectInfo *taskqueue.BuildIdRedirectInfo,
+) error {
+	assignedBuildId := ms.GetAssignedBuildId()
+	if !versioningStamp.GetUseVersioning() || versioningStamp.GetBuildId() == assignedBuildId {
+		// dispatch build ID is the same as wf assigned build ID, hence noop.
+		return nil
+	}
+
+	if (assignedBuildId != "" || ms.HasCompletedAnyWorkflowTask()) &&
+		(redirectInfo == nil || redirectInfo.GetAssignedBuildId() != assignedBuildId) {
+		// Workflow is already assigned to a build ID (or completed tasks by unversioned workers) but no redirect
+		// or a redirect based on a wrong assigned build ID is reported. This must be a task
+		// backlogged on an old build ID. rejecting this task, there should be another task scheduled on
+		// the right build ID.
+		return serviceerrors.NewInvalidDispatchBuildId()
+	}
+
+	// redirect is valid, updating mutable state and re-scheduling pending tasks which are not started yet.
+	newBuildId := versioningStamp.GetBuildId()
+	err := ms.UpdateBuildIdAssignment(newBuildId)
+	if err != nil {
+		return err
+	}
+
+	if ms.HasPendingWorkflowTask() && !ms.HasStartedWorkflowTask() &&
+		ms.GetPendingWorkflowTask().ScheduledEventID != startingTaskScheduledEventId &&
+		// speculative tasks do not go to matching
```

Comment:
WDYM don't go to matching?

Reply:
That comment is wrong (converted it to a TODO). Speculative tasks are not within scope for pre-release. Will add support later.